File: app/api/schemas/qlib_research.py
from __future__ import annotations

from datetime import date
from typing import Any, Optional

from pydantic import BaseModel, Field


class QlibStatusOut(BaseModel):
    status: str
    provider_uri: str
    universe: str
    freq: str
    package: str
    package_available: bool
    package_version: Optional[str] = None
    data_available: bool
    checks: dict[str, bool]
    notes: list[str]


class RunFactorMiningIn(BaseModel):
    provider_uri: Optional[str] = None
    universe: str = "csi300"
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    horizon: int = Field(default=1, ge=1, le=60)
    quantiles: int = Field(default=5, ge=2, le=20)
    top_k: int = Field(default=20, ge=1, le=200)
    freq: str = "day"
    factor_pool: Optional[list[str]] = None
    factor_limit: Optional[int] = Field(default=None, ge=1, le=500)


class FactorMiningRunOut(BaseModel):
    run_id: str
    status: str
    factor_count: int
    date_range: dict[str, Any]
    artifact_path: str
    top_factors: list[dict[str, Any]]
    quality_gate: Optional[dict[str, Any]] = None
    promotion_status: Optional[str] = None
    not_executable: Optional[bool] = None
    quality_reason_codes: list[str] = []


class BuildPortfolioIn(BaseModel):
    mining_run_id: str
    selected_factors: Optional[list[str]] = None
    weighting_method: str = "equal"
    top_n: int = Field(default=5, ge=1, le=50)
    long_top_n: int = Field(default=30, ge=1, le=500)


class PortfolioOut(BaseModel):
    portfolio_id: str
    created_at: str
    mining_run_id: str
    provider_uri: Optional[str] = None
    universe: Optional[str] = None
    selected_factors: list[str]
    weighting_method: str
    weights: dict[str, float]
    signal_artifact_path: str
    signal_count: int
    date_count: int
    timing_note: str
    quality_gate: Optional[dict[str, Any]] = None
    promotion_status: Optional[str] = None
    not_executable: bool = False
    quality_reason_codes: list[str] = []

File: web/src/types/qlib-research.ts
export type QlibStatus = {
  status: "READY" | "QLIB_NOT_READY" | "DATA_NOT_READY" | string;
  provider_uri: string;
  universe: string;
  freq: string;
  package: string;
  package_available: boolean;
  package_version?: string | null;
  data_available: boolean;
  checks: Record<string, boolean>;
  notes: string[];
};

export type QlibFactorMiningRun = {
  run_id: string;
  status: string;
  factor_count: number;
  date_range: Record<string, unknown>;
  artifact_path: string;
  top_factors: Record<string, unknown>[];
  generated_at?: string;
  universe?: string;
  provider_uri?: string;
  quality_gate?: ResearchQualitySummary | null;
  promotion_status?: string | null;
  not_executable?: boolean | null;
  quality_reason_codes?: string[];
};

export type QlibPortfolio = {
  portfolio_id: string;
  created_at: string;
  mining_run_id: string;
  provider_uri?: string | null;
  universe?: string | null;
  selected_factors: string[];
  weighting_method: string;
  weights: Record<string, number>;
  signal_artifact_path: string;
  signal_count: number;
  date_count: number;
  timing_note: string;
  quality_gate?: ResearchQualitySummary | null;
  promotion_status?: string | null;
  not_executable?: boolean;
  quality_reason_codes?: string[];
};

export type ResearchQualityCheck = {
  check_id: string;
  status: "PASS" | "WARN" | "FAIL" | "NOT_TESTED" | string;
  reason_code: string;
  message: string;
  evidence?: Record<string, unknown>;
  threshold?: Record<string, unknown>;
};

export type ResearchQualitySummary = {
  source_run_id?: string;
  quality_status?: "PASS" | "WARN" | "FAIL" | string;
  quality_score?: number;
  promotion_status?: string;
  reason_codes?: string[];
  research_ops_object_id?: string | null;
  artifact_path?: string;
  factor_status?: Record<string, string>;
};

export type ResearchQualityReport = ResearchQualitySummary & {
  source_type: string;
  evaluated_at: string;
  asof_date?: string;
  not_executable: boolean;
  checks: ResearchQualityCheck[];
  thresholds: Record<string, unknown>;
  findings_path?: string;
  source_artifact_path?: string;
  timing_note?: string;
};

export type QlibMiningPayload = {
  provider_uri?: string | null;
  universe: string;
  start_date?: string | null;
  end_date?: string | null;
  horizon: number;
  quantiles: number;
  top_k: number;
  freq?: string;
  factor_pool?: string[] | null;
  factor_limit?: number | null;
};

export type BuildQlibPortfolioPayload = {
  mining_run_id: string;
  selected_factors?: string[] | null;
  weighting_method: "equal" | "ic_weighted" | "rank_ic_weighted" | string;
  top_n: number;
  long_top_n: number;
};
